### ISSUE TITLE
Alternate proposal for `lookup` function of the new `Record` module

### ISSUE THREAD

**Issue by joshburgess** [gcanti/fp-ts#608]
**Do you want to request a _feature_ or report a _bug_?**

Alternate proposal for `lookup` function of the new `Record` module

**What is the current behavior?**

The current definition of the `lookup` function in the new `Record` module seems strange to me. 

```
/**
 * Lookup the value for a key in a dictionary
 * @since 1.10.0
 */
export const lookup = <A>(key: string, fa: { [key: string]: A }): Option<A> => {
  return fa.hasOwnProperty(key) ? some(fa[key]) : none
}
```

When attempting to use this, I ran into 2 problems:

1. TypeScript was not able to properly infer the generic A on it's own, and thus not able to properly infer return type. The type of the return value was inferred as `Option<{}>`

2. When attempting to provide the type for the generic manually, I realized this was not going to work for me, because the type `{ [key: string]: A }` restricts `fa` key-value pairs to be uniform (All keys in fa must be of type `string` and all values of `fa` must be of the type you provide for the generic `A`).

This doesn't seem right to me. Is that the idea of this module? Record.ts is only for singleton key-val pairs or Records with uniform key-val pairs?

I wrote my own version of the function which is able to infer everything, including the actual value associated with the key you are performing the lookup on, without needing to manually pass in a type for the generic:

```
export const lookup = <
  A extends Record<PropertyKey, unknown>,
  B extends keyof A
>(
  key: B,
  fa: A,
): Option<A[B]> => (fa.hasOwnProperty(key) ? some(fa[key]) : none)
```

I used `PropertyKey` there, because I don't see the need to restrict the Record to only `string` keys, but that part isn't important.

It's possible I just misunderstood the purpose of the `Record` module, but I'm not sure.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem via https://codesandbox.io/ or similar.**

**What is the expected behavior?**

I would expect to not need to pass a generic in when it's possible to define this function in such a way that it gets inferred.

**Which versions of fp-ts, and which browser and OS are affected by this issue? Did this work in previous versions of fp-ts?**

This is in the most recent release with the new `Record` module.

**Comment by joshburgess** [comment]
Hmmm, after looking over the module a little more, I see that all of the other functions also have the requirement of the values being of the same type. I suppose this was to allow implementing some of the typeclasses.

I guess that leads me to 2 questions then:

1. Can we add some documentation somewhere describing the differences between `StrMap` and `Record`?

and

2. Is there room for more generic utility functions without such restrictions, like the one I posted above, anywhere in the library?

**Comment by gcanti** [comment]
> Record.ts is only for singleton key-val pairs or Records with uniform key-val pairs?

Yes, it's for handling `Record<string, A>` (= `{ [key: string]: A }`), i.e. a non wrapped `StrMap`

> Can we add some documentation somewhere describing the differences between StrMap and Record?

Sure

> Is there room for more generic utility functions without such restrictions, like the one I posted above, anywhere in the library?

Maybe, what's the use case of

```
export const lookup = <A extends Record<PropertyKey, unknown>, B extends keyof A>(key: B, fa: A): Option<A[B]> =>
  fa.hasOwnProperty(key) ? some(fa[key]) : none
```
